src/lock.h

```diff
@@ -1,55 +1,54 @@
 #ifndef ZIM_LOCK_H
 #define ZIM_LOCK_H
 
-#include <memory>
+#include <algorithm>
 #include <mutex>
+#include <system_error>
+#include <vector>
 
 class MultiMutex {
   public:
-    explicit MultiMutex(std::recursive_mutex* m1)
-     : m1(m1), m2(nullptr) {}
-    MultiMutex(std::recursive_mutex* m1, std::unique_ptr<MultiMutex> m2)
-     : m1(m1), m2(std::move(m2)) {}
-
-    template<typename Iterator>
-    static std::unique_ptr<MultiMutex> create(Iterator begin, Iterator end) {
-      if (begin == end) {
-        return nullptr;
-      }
-
-      auto second = begin;
-      second++;
-      if (second == end) {
-        return std::make_unique<MultiMutex>(*begin);
-      }
-
-      return std::make_unique<MultiMutex>(*begin, MultiMutex::create(second, end));
-    }
+    explicit MultiMutex()
+     : m_mutexes()
+     {}
+    explicit MultiMutex(const std::vector<std::recursive_mutex*> mutexes)
+     : m_mutexes(mutexes) {
+    // By sorting the mutex, we avoid the simple case when 3 concurrent multi lock:
+    // - (A, B)
+    // - (B, C)
+    // - (C, A)
+    // As we sort, we will have :
+    // - (A, B)
+    // - (B, C)
+    // - (A, C)
+    //
+    // And no deadlock can occurs.
+       std::sort(m_mutexes.begin(), m_mutexes.end());
+     }
 
     void lock() {
-      if (m2) {
-        std::lock(*m1, *m2);
-      } else {
-        m1->lock();
+      auto current_locked = 0;
+      for (auto mutex:m_mutexes) {
+        try {
+          mutex->lock();
+          current_locked += 1;
+        } catch(const std::system_error& e) {
+          while (current_locked) {
+            m_mutexes[--current_locked]->unlock();
+          }
```

Comment:
Please extract this into a private member function (e.g. `unwindLocks(size_t n)`) and reuse it in `unlock()` (as `unwindLocks(m_mutexes.size())`). I feel more comfortable if mutexes are unlocked in the reverse order of locking.